### PR TITLE
Block validator

### PR DESCRIPTION
**Pull request by cintel**
Enhance block validator design, add block validation of completeness and signature. Enhance test framework to support this additional validation. 
Add BlockValidator test for various fork conditions to resolve. 
Clean up callback interface between BlockValidator and ChainController

### CODE REVIEW

File: validator/sawtooth_validator/journal/chain.py [hunk 1]
```diff
@@ -62,10 +64,28 @@ def __init__(self, consensus,
         }
 
     def _is_block_complete(self, blkw):
+        """
```
Comment (agunde406) [inline validator/sawtooth_validator/journal/chain.py]: The completer will only pass block to the journal if they are deem complete. What is your reasoning for "re-checking" the completeness?

**Comment by agunde406** [comment]
The "fixes" and "clean up" commits should be squashed into the corresponding commits.

File: validator/sawtooth_validator/journal/chain.py [hunk 2]
```diff
@@ -43,161 +54,246 @@ def __init__(self, consensus,
...
+            'new_chain': [],
+            'cur_chain': [],
+            'committed_batches': [],
+            'uncommitted_batches': [],
+        }
 
-    @property
-    def chain_head(self):
-        return self._chain_head
+    def _is_block_complete(self, blkw):
```
Comment (agunde406) [inline validator/sawtooth_validator/journal/chain.py]: The completer will only pass block to the journal if they are deem complete. What is your reasoning for "re-checking" the completeness?

Comment (cintel) [inline validator/sawtooth_validator/journal/chain.py]: 1) The completer is not the only source of blocks. the blocks can be loaded from the BlockCache or the BlockStore. Both of risk of having corrupted data. Long term it is not clear the policy of the BlockCache will be only to contain completed blocks. 
2) During testing blocks do not go thru the completer. Best to fail early on bad data.

**Pull request merged**